sql: add validation to cascading zone configs

ridwanmsharif opened the pull request:
I found a few hours free and so took a second stab at https://github.com/cockroachdb/cockroach/pull/32059. Thought it'd be fun and new to try adding something to SQL for this, so @knz please feel free to tell me if what I've done here makes no sense. 

There are to be two changes to be made: 
1) Allow for the `COPY FROM PARENT` command so users can do something like:
` ALTER ... CONFIGURE ZONE USING range_min_bytes = ..., range_max_bytes = COPY FROM PARENT`

2) Add validation to the zone configs so certain fields are written in tandem. The fields are:
      - `NumReplicas` and `Constraints` (Only when per-replica constraints are in use)
      - `RangeMinBytes` and `RangeMaxBytes`
      - `LeasePreferences` and `Constraints`

Each of the changes are contained in their own commits.

Commit:
sql: add COPY FROM PARENT command for zone configs

Adds a `COPY FROM PARENT` command to zone configs to
solidify a field. Solidifying a field means, the field will
now be explicitly set to the value that applied to it.
Any change to a parent zone on the field will no longer
affect it.
This change will allow people to do the following:

`ALTER .... CONFIGURE ZONE USING range_min_bytes = ...,
range_max_bytes = COPY FROM PARENT` instead of figuring
out what the `RangeMaxBytes`should be and populating it manually.
This change will allow us to add a validation constraint
that tandem fields are set together (for more details see
the cascading zone config RFC).

Release note (sql change): New SQL command added.
`ALTER ... CONFIGURE ZONE USING field_name = COPY FROM PARENT`
now solidifies/explicitly sets the field.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32861)
<!-- Reviewable:end -->

Commit:
config: Add zone validation for tandem fields

As described in the cascading zone RFC, a few additional
constraints are required on zone config usage so no zone
is ever in an invalid state if their parent zone changes
a value.

Release note: None

knz reviewed (commented):
Code LGTM. Well done. A few comments below.

> Reviewed 10 of 10 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-:-LT2S8eCDU8BrkHU8Y8G:b-tnjqkw)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/split.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT2Rdw-AU4EXmvanTF1:-LT2Rdw00i_WijS2WGec:b-fmxt7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d3f0ede204d8affb65c010bd2cfa2044308f55/pkg/cmd/roachtest/split.go#L143)):*
> ```Go
> 		setRangeMaxBytes := func(maxBytes int) {
> 			stmtZone := fmt.Sprintf(
> 				"ALTER RANGE default CONFIGURE ZONE USING range_max_bytes = %d, range_min_bytes = COPY FROM PARENT",
> ```

please check with the powers that be that this test code is only used with 2.2/master; if it is also used with 2.1 you'll need to backport the syntax change to 2.1 before you can merge the test update.
___
*[pkg/sql/set_zone_config.go, line 472 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT2Rv2P2-9bmMS5pwsI:-LT2Rv2P2-9bmMS5pwsJ:bq7dlwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d3f0ede204d8affb65c010bd2cfa2044308f55/pkg/sql/set_zone_config.go#L472)):*
> ```Go
> 	// Per-replica constraints cannot be set unless num_replicas is explicitly set
> 	if err := zoneToWrite.ValidateTandemFields(); err != nil {
> 		return pgerror.NewErrorf(pgerror.CodeInternalError,
> ```

This is not an internal error. Internal errors are bugs in CockroachDB; also internal errors get uploaded to our telemetry data.

Here you are simply reporting an error to the client/user. You can use one of the other error codes instead.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32861)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/parser/sql.y, line 1416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT2SL-RDbBMfsIKhe7F:-LT2SL-RDbBMfsIKhe7G:b-kcas4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bd4344dab2d75575c72644c8052031663329360/pkg/sql/parser/sql.y#L1416)):*
> ```Yacc
> 
> var_set_list:
>   var_name '=' COPY FROM PARENT
> ```

Also add the missing tests to `parse_test.go`.


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32861)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/split.go, line 143 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT2Rdw-AU4EXmvanTF1:-LT3_Efk2aaLGmSUXbdB:b-abr05s) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d3f0ede204d8affb65c010bd2cfa2044308f55/pkg/cmd/roachtest/split.go#L143)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

please check with the powers that be that this test code is only used with 2.2/master; if it is also used with 2.1 you'll need to backport the syntax change to 2.1 before you can merge the test update.
</blockquote></details>

This test uses 2.2 but good note, changes the other one to use the min bytes explicitly.
___
*[pkg/sql/set_zone_config.go, line 472 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT2Rv2P2-9bmMS5pwsI:-LT3_N3J1wk6S-ZcqsLp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25d3f0ede204d8affb65c010bd2cfa2044308f55/pkg/sql/set_zone_config.go#L472)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This is not an internal error. Internal errors are bugs in CockroachDB; also internal errors get uploaded to our telemetry data.

Here you are simply reporting an error to the client/user. You can use one of the other error codes instead.
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 1416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT2SL-RDbBMfsIKhe7F:-LT3_Ng39_7-c0uTfkrq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bd4344dab2d75575c72644c8052031663329360/pkg/sql/parser/sql.y#L1416)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Also add the missing tests to `parse_test.go`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

benesch commented:
Approach looks fantastic to me! If @knz is happy, I'm happy.

Commit:
sql: Update the zone specifier formatter

The formatter is updated to now recognise the COPY FROM PARENT
command when the value in the KVOption is nil.

Release note: None

knz commented on pkg/sql/sem/tree/zone.go:
```diff
@@ -102,6 +102,18 @@ func (node *SetZoneConfig) Format(ctx *FmtCtx) {
-		ctx.FormatNode(&node.Options)
+		kvOptions := node.Options
+		for i, kv := range kvOptions {
+			if i > 0 {
```
```go
comma := ""
for ... {
   WriteString(comma)
   ...
   comma = ", "
}
```

knz commented on pkg/sql/set_zone_config.go:
```diff
@@ -210,6 +210,10 @@ func (n *setZoneConfigNode) startExec(params runParams) error {
+				if optionStr.Len() > 0 {
+					optionStr.WriteString(", ")
+				}
+				fmt.Fprintf(&optionStr, "%s = COPY FROM PARENT", string(*name))
```
wait what no you should not convert the name to a string!!

it's `fmt("%s = ...", name)`

knz reviewed (commented):
> Reviewed 7 of 7 files at r3, 5 of 5 files at r4, 1 of 3 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32861)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/set_zone_config.go, line 473 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9t5BD3DUFwJzO4Mb6:-LT9t5BD3DUFwJzO4Mb7:b-4bhp4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1df37d0aa6a0e8a721c9589be1cd716f92ffff/pkg/sql/set_zone_config.go#L473)):*
> ```Go
> 	if err := zoneToWrite.ValidateTandemFields(); err != nil {
> 		return pgerror.NewErrorf(pgerror.CodeInvalidParameterValueError,
> 			fmt.Sprintf("could not validate zone config: %s", err)).SetHintf(
> ```

1) we use `%v` for errors not `%s`

2) `NewErrorf` means you can use formatting -- so it's `pgerror.NewErrorf(code, "foo %v", err)` not `NewErrorf(code, fmt.Sprintf(...))`
___
*[pkg/sql/set_zone_config.go, line 134 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tTMt-2J3nqbLI3RT:-LT9tTMt-2J3nqbLI3RU:bxdcx4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L134)):*
> ```Go
> 		for _, opt := range n.Options {
> 			if _, alreadyExists := options[opt.Key]; alreadyExists {
> 				return nil, fmt.Errorf("duplicate zone config parameter: %q", tree.ErrString(&opt.Key))
> ```

while you're at it please convert the errors around here to pgerrors
___
*[pkg/sql/set_zone_config.go, line 216 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tZb16gqzniycXQFY:-LT9tZb16gqzniycXQFZ:buq9921) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L216)):*
> ```Go
> 					optionStr.WriteString(", ")
> 				}
> 				fmt.Fprintf(&optionStr, "%s = COPY FROM PARENT", *name)
> ```

pass `name` not `*name`, they are not the same.
___
*[pkg/sql/set_zone_config.go, line 224 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tdqz5SZzT0jAg2Tk:-LT9tdqz5SZzT0jAg2Tl:bf8dho6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L224)):*
> ```Go
> 			}
> 			if datum == tree.DNull {
> 				return fmt.Errorf("unsupported NULL value for %q", tree.ErrString(name))
> ```

ditto pgerror
___
*[pkg/sql/set_zone_config.go, line 231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tctp7DZUTF4xWizd:-LT9tctp7DZUTF4xWize:b-4a9jew) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L231)):*
> ```Go
> 				optionStr.WriteString(", ")
> 			}
> 			fmt.Fprintf(&optionStr, "%s = %s", *name, datum)
> ```

ditto `name` vs `*name`


<!-- Sent from Reviewable.io -->

ridwanmsharif reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-:-LTYAtshEclB4BlrgDD8:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/set_zone_config.go, line 473 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9t5BD3DUFwJzO4Mb6:-LTYArpo1FJ6REcAzSg9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e1df37d0aa6a0e8a721c9589be1cd716f92ffff/pkg/sql/set_zone_config.go#L473)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) we use `%v` for errors not `%s`

2) `NewErrorf` means you can use formatting -- so it's `pgerror.NewErrorf(code, "foo %v", err)` not `NewErrorf(code, fmt.Sprintf(...))`


</blockquote></details>

Done.
___
*[pkg/sql/set_zone_config.go, line 134 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tTMt-2J3nqbLI3RT:-LTYAmwQ7-JizcpF6gEw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L134)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

while you're at it please convert the errors around here to pgerrors
</blockquote></details>

Done.
___
*[pkg/sql/set_zone_config.go, line 216 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tZb16gqzniycXQFY:-LTYAoWf93NBBhsUqc0x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L216)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

pass `name` not `*name`, they are not the same.
</blockquote></details>

Done.
___
*[pkg/sql/set_zone_config.go, line 224 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tdqz5SZzT0jAg2Tk:-LTYApAQ3IfNecd_BcXG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L224)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto pgerror
</blockquote></details>

Done.
___
*[pkg/sql/set_zone_config.go, line 231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-LT9tctp7DZUTF4xWizd:-LTYAq8v1HxEwiE2FMqi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a13ac50fa3cea066d6fd947ece96290e6d5b342d/pkg/sql/set_zone_config.go#L231)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto `name` vs `*name`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 9 files at r7, 4 of 5 files at r8, 3 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32861#-:-LTZ0aMF7ii1JosLA8l2:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

ridwanmsharif commented:
bors r+
